src/vcSHP.cpp

```diff
@@ -0,0 +1,741 @@
+      for (int i = 0; i < record.data.polyM.pointsCount; i++)
+        readPosition = UnpackDouble(pSHP->localBigEndian, readPosition, &record.data.polyM.MValue[i], buffer);
+    }
+  }
+  break;
+  case vcSHP_Type::MultiPointM:
+  {
+    readPosition = UnpackDouble(pSHP->localBigEndian, readPosition, &record.data.multiPointM.MBRmin.x, buffer);
+    readPosition = UnpackDouble(pSHP->localBigEndian, readPosition, &record.data.multiPointM.MBRmin.y, buffer);
+    readPosition = UnpackDouble(pSHP->localBigEndian, readPosition, &record.data.multiPointM.MBRmax.x, buffer);
+    readPosition = UnpackDouble(pSHP->localBigEndian, readPosition, &record.data.multiPointM.MBRmax.y, buffer);
+
+    readPosition = UnpackInt32(pSHP->localBigEndian, readPosition, &record.data.multiPointM.pointsCount, buffer);
+    record.data.multiPointM.points = udAllocType(udDouble3, record.data.multiPointM.pointsCount, udAF_Zero);
+    for (int i = 0; i < record.data.multiPointM.pointsCount; i++)
+    {
+      udDouble3 *p = &record.data.multiPointM.points[i];
+      readPosition = UnpackDouble(pSHP->localBigEndian, readPosition, &p->x, buffer);
+      readPosition = UnpackDouble(pSHP->localBigEndian, readPosition, &p->y, buffer);
+      p->z = 0;
+    }
+    // optional:
+    record.data.multiPointM.MProvided = false;
+    leftBytes = record.length - readPosition;
+    if (leftBytes == 8 * (2 + record.data.multiPointM.pointsCount))
+    {
+      record.data.multiPointM.MProvided = true;
+      readPosition = UnpackDouble(pSHP->localBigEndian, readPosition, &record.data.multiPointM.MRange.x, buffer);
+      readPosition = UnpackDouble(pSHP->localBigEndian, readPosition, &record.data.multiPointM.MRange.y, buffer);
+
+      record.data.multiPointM.MValue = udAllocType(double, record.data.multiPointM.pointsCount, udAF_Zero);
+      for (int i = 0; i < record.data.multiPointM.pointsCount; i++)
+        readPosition = UnpackDouble(pSHP->localBigEndian, readPosition, &record.data.multiPointM.MValue[i], buffer);
+    }
+  }
+  break;
+  default:
+  break;
+  }
+
+  pSHP->shpRecords.PushBack(record);
+  offset += readPosition;
+
+epilogue:
+  if (buffer)
+    udFree(buffer);
+
+  if (result != udR_Success)
+    vcSHP_ReleaseRecord(record);
+
+  return result;
+}
+
+udResult vcSHP_LoadShpFile(vcSHP *pSHP, const char *pFilename)
+{
+  udResult result;
+
+  udFile *pFile = nullptr;
+  uint8_t *headerBuffer = nullptr;
+  uint32_t offset = 0;
+
+  vcSHP_Header &header = pSHP->shpHeader;
+  UD_ERROR_CHECK(udFile_Open(&pFile, pFilename, udFOF_Read));
+
+  headerBuffer = udAllocType(uint8_t, 100, udAF_Zero);
+  UD_ERROR_CHECK(udFile_Read(pFile, headerBuffer, 100 * sizeof(uint8_t)));
+  // 00 00 27 0a
+  UD_ERROR_IF((headerBuffer[0] != 0 || headerBuffer[1] != 0 || headerBuffer[2] != 0x27 || headerBuffer[3] != 0x0a), udR_ParseError);
+
+  header.fileLength = (headerBuffer[24] << 24) | (headerBuffer[25] << 16) | (headerBuffer[26] << 8) | headerBuffer[27];
+  if (header.fileLength < UINT_MAX / 2)
```

Comment:
You should simplify this to `header.fileLength = 2 * udMin(header.fileLength, UINT32_MAX / 2);`. You should also use `UINT32_MAX` because it can differ to `UINT_MAX`, since `int` only needs to be at least 16bit.

Reply:
fixed.